Fix beginless and endless range comparisons

AlexMooney opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because some endless and beginless range comparisons raise `RangeError` due to `ActiveSupport::CompareWithRange`

### Detail

This Pull Request changes the `CompareWithRange#===` and `CompareWithRange#include?` methods to support begin and endless ranges on the left and right hand side of the comparison.

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

POR:
```ruby
> (1..) === (2..)
=> false
> (..1) === (..2)
=> false
```
Rails 7.1.3:
```ruby
> (1..) === (2..)
RangeError: cannot get the last element of endless range
from …/activesupport-7.1.3/lib/active_support/core_ext/range/compare_range.rb:23:in `last'
> (..1) === (..2)
RangeError: cannot get the first element of beginless range
from …/activesupport-7.1.3/lib/active_support/core_ext/range/compare_range.rb:24:in `first'
```
This PR:
```ruby
> (1..) === (2..)
=> true
> (2..) === (1..)
=> false
> (..1) === (..2)
=> false
> (..2) === (..1)
=> true
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

AlexMooney committed:
Correct beginless and endless range comparisons

Some range comparisons raise `RangeError`, for example:
```ruby
> (1..) === (2..)
RangeError: cannot get the last element of endless range
from …/activesupport-7.1.3/lib/active_support/core_ext/range/compare_range.rb:23:in `last'
```

skipkayhil reviewed (commented):
Both methods have this comment which should be changed:
```
# The given range must be fully bounded, with both start and end.
```

---

Thinking out loud: It seems like `include?` ends up very similar if not the same as `cover?`, are there any differences left?

Edit: introduced in 99c64829ce71ecf07b576960ea8ff01dfee61be7, I wonder if we should just recommend people use `cover?` now 🤔

yawboakye reviewed (commented):
@AlexMooney both `===` and `include?` method definitions have comments that say that the range should be bounded. if we're relaxing that requirement, it might be a good idea to remove those comments too?

yawboakye reviewed (commented):
@AlexMooney i've looked at the code over and over again and i think it could be written in a clearer way. a lot of things get in the way of comprehension. i recommend that you separate the endless/beginless range concern, and handle with a private method.

AlexMooney commented:
> Both methods have this comment which should be changed:
> 
> ```
> # The given range must be fully bounded, with both start and end.
> ```

Thanks for catching that! I've removed it.

> Thinking out loud: It seems like `include?` ends up very similar if not the same as `cover?`, are there any differences left?

I don't think there are any differences in numerical values, but in non-numeric ranges you can get [different answers](https://rubyapi.org/3.1/o/range#method-i-member-3F):
```ruby
> ('a'..'d').=== 'cc'
true
> ('a'..'d').include? 'cc'
false
```

> Edit: introduced in [99c6482](https://github.com/rails/rails/commit/99c64829ce71ecf07b576960ea8ff01dfee61be7), I wonder if we should just recommend people use `cover?` now 🤔

The reason I ended up here is this extension messing with ranges makes some spec like `expect(array_of_beginless_ranges).to contain_exactly(expected)` blow up.  It might be better to remove these overrides entirely in some breaking Rails release, though!

> @AlexMooney i've looked at the code over and over again and i think it could be written in a clearer way. a lot of things get in the way of comprehension. i recommend that you separate the endless/beginless range concern, and handle with a private method.

@yawboakye I've rewritten it to use smaller private methods so that I could name the concepts involved.  Let me know if you like this way better.

yawboakye commented on activesupport/lib/active_support/core_ext/range/compare_range.rb:
```diff
@@ -36,21 +24,44 @@ def ===(value)
+
+    private
+      def compare_with_range(other, &block)
+        if other.is_a?(::Range)
```
let's move this condition outside. `compare_with_range` should only ever be called if `other` is indeed a range. also we could make `include?` an alias of `===`. they're essentially the same. in the end we should have something like this:

```ruby
def ===(other)
  return compare_with_range(other) if other.is_a?(::Range)

  super(other)
end

alias :include? :===
```

yawboakye reviewed (commented):
one more (and final) refactoring. from where i sit, both `===` and `include?` seem to have identical implementations, so i recommend we make one the alias of the other.

AlexMooney commented:
> one more (and final) refactoring. from where i sit, both `===` and `include?` seem to have identical implementations, so i recommend we make one the alias of the other. 

That is almost but not entirely so. The `super` that they yield is subtly different from one another. See the string range from my previous comment for an example of the difference.

yawboakye commented:
> That is almost but not entirely so. The `super` that they yield is subtly different from one another. See the string range from my previous comment for an example of the difference.

good point 👍! we can still move the range check condition outside of `compare_with_range` though so that we’re only ever entering that method when we’re certain `other` is a range.

yawboakye commented:
looks good to me cc @skipkayhil

AlexMooney commented:
@skipkayhil can I get another review, please?
